xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/viewers/attachments.js

```diff
@@ -233,11 +230,45 @@ require(['jquery', 'xwiki-events-bridge'], function($) {
     })
   });
   /**
-   * Updating the number of files in AttachmentsTab and in More actions menu.
+   * Make sure the number of attachments displayed matches after the livetable has been refreshed, as is the case after
+   * attachment upload.  
+   */
+  $(document).on("xwiki:livetable:docAttachments:loadingComplete", function() {
+    updateCount(docAttachmentsLivetable.totalRows);
+  })
+  /**
+   * On delete action, show a confirmation modal and save the element that triggered this event to be able to access
+   * information after confirmation.
+   */
+  $(document).on('click', '.attachmentActions .actiondelete', function(event) {
+    event.preventDefault();
+    var modal = $(event.currentTarget).closest('.xwiki-livetable-container').next('.deleteAttachment');
+    modal.data('relatedTarget', event.currentTarget);
+    modal.modal('show');
+  });
+  /**
+   * Update the locations that display the attachments count with the new number or, when it is not provided, make
+   * another request to get it.
    */
-  var updateCount = function() {
+  var updateCount = function(attachmentsNumber) {
+    if (attachmentsNumber && attachmentsNumber >= 0) {
+      updateAttachmentsNumber(attachmentsNumber);
+    } else {
+      $.ajax({
+        url: XWiki.currentDocument.getURL('get', 'xpage=xpart&vm=attachmentsjson.vm'),
+        success: function(data) {
+          updateAttachmentsNumber(data.totalrows);
+        }
+      });
+    }
+  };
+  /**
+   * Update the number of attachments from AttachmentsTab and More actions menu.
+   *
+   * @param attachmentsNumber the total number of attachments 
+   */
+  var updateAttachmentsNumber = function(attachmentsNumber) {
     var itemCount = $('#Attachmentstab').find('.itemCount');
-    var attachmentsNumber = $("#Attachmentspane .attachment").length;
     if(itemCount) {
       itemCount.text(l10n['docextra.extranb'].replace("__number__", attachmentsNumber));
     };
```

Comment:
Unnecessary semicolon (not your change, but same function). Also, I get a complaint about the duplicate selectors below in Intellij IDEA.